src/Microsoft.Health.Fhir.Liquid.Converter/Utilities/ComplexObjectFilterUtility.cs

```diff
@@ -0,0 +1,173 @@
+// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.Health.Fhir.Liquid.Converter.Utilities
+{
+    public static class ComplexObjectFilterUtility
+    {
+        private static readonly char[] Delimeters = new[] { '.', '[', ']' };
+
+        public static object[] Select(object[] input, string path, string value = null)
+        {
+            // Split path on . and [], [5]
+            Queue<string> pathKeys = SplitObjectPath(path);
+
+            List<object> ret = new List<object>();
+
+            foreach (object obj in input)
+            {
+                // Clone the queue to scope the path at this level for each loop
+                var localPath = new Queue<string>(pathKeys);
+                if (ObjHasValueAtPath(obj, localPath, value))
+                {
+                    // This object has our value at the path so add it to our return
+                    ret.Add(obj);
+                }
+            }
+
+            return ret.ToArray<object>();
+        }
+
+        public static bool ObjHasValueAtPath(object input, Queue<string> path, string value)
+        {
+            // If we're at the end of the path then check the value
+            if (path.Count == 0)
+            {
+                // Return true if value is null and an equality check otherwise
+                return value == null || input.ToString().Equals(value);
+            }
+
+            // Get our key name
+            var key = path.Dequeue();
+
+            // Check if our key is an object property or an array reference
+            if (key.Contains('['))
+            {
+                // Our required path dictates we should have an array and if not
+                // then this is not a match
+                if (input is object[] inputArray)
```

Comment:
To be even more generic you could do `input is IEnumerable inputEnumerable`.

Reply:
Does an Enumerable allow access with brackets such as `inputEnumerable[3]`?

Reply:
I think the correct type would actually be `IEnumerable<object>` and if you import `System.Linq` you can use the `.GetElementAt(index)` property, which behaves the same as an index.

I did think of another problem though, do you have a test case that covers selecting from an array of value types? For example with this json:
```json
{
  "numbers": [ 0, 1, 2, 3, 4 ]
}
```
And this path:
```
numbers[2]
```
I don't think you can convert `int[]` to `object[]` so I would expect this example to fail.

Reply:
Hmm, I'd need to change the code and add a include to make this happen and I'm personally not a fan of `.GetElementAt` over `[]` so I think I'll let this suggestion slide. Thank you.